<a href="https://colab.research.google.com/github/SuMoralesO/Business-Analytics/blob/main/1_Reto_1_Modelo_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 0. Se procede con la carga de las librerías.


In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

#1. Se procede con la carga de los datos.

In [ ]:
Data = "/content/drive/MyDrive/Analítica de Negocios/1. BD2_CarCrash.xlsx"
Datos = pd.read_excel(Data,sheet_name = 0)
Datos.dropna()
Datos.head()

# Creamos el DataMart.

Info = Datos[["Weekday", "ViolCat", "ClearWeather","Month", "Daylight"]]
yd = Datos[["Highway"]]

# 2. Se procede con la configuración y evaluación del modelo Naive Bayes (Highway).

In [ ]:
mnb = GaussianNB()
mnb.fit(Info,yd)

# Analizamos los resultados arrojados por el modelo.
u = mnb.theta_  # Son las medias.
print("Las medias de las variables son: \n", u)
sigma = np.sqrt(mnb.var_)
print("Los valores de las desviaciones son: \n", sigma)

# Se procede con el análisis de las variaciones de las variables.
LS = u + sigma
print("Los límites superiores para las variables son: \n", LS)
LI = u - sigma
print("Los límites inferiores para las variables son: \n", LI)

# Prediciento los valores de preaprobación del valor deseado.
ydp = mnb.predict(Info)

# Se procede con la evaluación del comportamiento del modelo.
cm = confusion_matrix(yd,ydp)
print("La matriz de confusión es: \n", cm)

VN = cm[0,0]; FP = cm[0,1]; FN = cm[1,0]; VP = cm[1,1]

Ex = (VP + VN)/len(Info)
print("La Exactitud del Modelo es: ", Ex)

Ter = (FP + FN)/len(Info)
print("La Tasa de Error del Modelo es: ", Ter)

Sen = VP/(VP + FN)
print("La Sensibilidad del Modelo es: ", Sen)

Esp = VN/(VN + FP)
print("La Especificidad del Modelo es: ", Esp)

Pres = VP/(VP + FP)
print("La Precisión del Modelo es: ", Pres)

VPN = VN/(VN + FN)
print("El valor de Predicción Negativo del Modelo es: ", VPN)

# Para evaluar un accidente.
XSC = [[7,1,1,3,0]]
ydsc = mnb.predict(XSC)
print("El accidente es: ", ydsc)

Las medias de las variables son: 
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
Los valores de las desviaciones son: 
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
Los límites superiores para las variables son: 
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
Los límites inferiores para las variables son: 
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es: 
 [[4560  325]
 [1563  257]]
La Exactitud del Modelo es:  0.7184190902311708
La Tasa de Error del Modelo es:  0.2815809097688292
La Sensibilidad del Modelo es:  0.1412087912087912
La Especificidad del Modelo es:  0.9334698055271239
La Precisión del Modelo es:  0.4415807560137457
El valor de Predicción Negativo del Modelo 

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


# 3. Se procede con la estimación.

In [ ]:
Datosn = pd.read_excel(Data,sheet_name = 1)

# Creamos el DataMart.
Infon = Datosn[["Weekday", "ViolCat", "ClearWeather","Month", "Daylight"]]
ydn = Datosn[["Highway"]]

# Hacer el pronóstico para todos los datos disponibles.
ydpn = mnb.predict(Infon)

# Se procede con la creación del informe.
InfoSalida = np.column_stack((Infon, ydpn, ydn))
df_sal = pd.DataFrame(InfoSalida)
df_sal.to_excel("Informe.xlsx")

#**Análisis de los datos**


La base de datos presenta un total de 6705 datos, de los cuales el 27.14% (1820) son accidentes que ocurrieron en una autopista y el 72.86% (4885) no ocurrieron en una autopista.

El modelo propuesto para el pronóstico de accidentes en una autopista tiene 5 variables que son: El día de la semana, el nivel de impacto, si hay un cielo despejado, el mes y la cantidad de luz diurna. Este modelo nos da como resultado una exactitud del 71.84%, indicando que el modelo tiende a equivocarse en un 28.16% que es su tasa de error. Con las métricas de sensibilidad, que arrojó un resultado del 14.12%, y especificidad, que arrojó un resultado del 93.35%, nos damos cuenta de que el modelo funciona mucho mejor al momento de identificar accidentes ocurridos en lugares distintos a la autopista. Además, la métrica de valor de predicción negativo indica que entre todos los resultados que arrojó como negativos, estuvo correcto en un 74.47% mientras que con la métrica de precisión sabemos que del total de resultados que arrojó como positivos estuvo correcto en un 44.16%.

Al momento de clasificar un accidente ocurrido durante un domingo del mes de marzo con un cielo despejado, sin luz diurna y con un nivel de impacto de 1, nos da como resultado que no ocurrió en la autopista.

A pesar de que la base de datos no presenta accidentes nuevos que aún no hayan sido clasificados según si ocurrieron en autopista o no, se realiza la estimación como método para comparar cuales datos clasificó correctamente.
